### ISSUE TITLE
Detecting failure to start

### ISSUE THREAD

**Issue by craigraw** [05nelsonm/kmp-tor#286]
I'm attempting to map `kmp-tor` to Sparrow's [TorService](https://github.com/sparrowwallet/sparrow/blob/ebd629db3a7e160ce2f55c9bdb8c235c15ba8a22/src/main/java/com/sparrowwallet/sparrow/net/TorService.java). TorService is a JavaFX `Service`, which is the recommended approach for a JavaFX application to use background threads to do work. There are only two possible outcomes from a JavaFX service - success or failure, with the [ability to register event handlers](https://github.com/sparrowwallet/sparrow/blob/f06b859c82a2fcc678f0b027b786266e5ba57342/src/main/java/com/sparrowwallet/sparrow/AppServices.java#L410-L438) for each situation. Failure is usually triggered by an exception in the `Task` in the `Service` doing the work.

Now, `kmp-tor` is already using background threads to do the work, notifying progress with event listeners, so this is a little more tricky. I can see how at least how a lock/signal could work once a success state has been reached (for example, `managerEventAddressInfo` has been called). However I'm unsure how to do this in general for a failure to start - would this be any call to `managerEventError`?

Side note: I was planning to simulate an error state by starting Tor externally on the same ports configured for the socks and control port. However, I found that in this case `kmp-tor` selects other unused ports and starts fine. I'm actually not unhappy with this behaviour - even though it differs from how Sparrow currently works, it's probably better to start Tor on another port than show an error in this case. But that raises a secondary question - how to determine the Tor control port in this situation - always read the file Tor writes, or is there a better way? Also, any ideas on another way to trigger failure to start?

**Comment by 05nelsonm** [comment]
> notifying progress with event listeners

Yes on some interactions with Tor via it's control port, tor only responds with success or failure and then dispatches a `6xx` coded event. For Java you can use the callback extension, as noted in the `README`, whereby most calls that actually return data (not just a `250 OK`) can be interacted with right there in the callback lambda.

As for event dispatched data (where tor only returns a `250 OK` and pushes an event), you can attach a listener as a "callback" to catch the event that tor dispatches and then remove it (strongly suggest a timeout detachment, depending on what the call is). 

All available interactions with Tor via it's control port can be viewed [HERE](https://github.com/05nelsonm/kmp-tor/tree/master/library/controller/kmp-tor-controller-common/src/commonMain/kotlin/io/matthewnelson/kmp/tor/controller/common/control/usecase)

> However I'm unsure how to do this in general for a failure to start - would this be any call to managerEventError

If you use the callback extension, you can call `torManager.start`. This will complete the entire start phase of tor up to when it begins bootstrapping. After bootstrapping, `AddressInfo` is dispatched with all of your configured ports/sockets.

> However, I found that in this case kmp-tor selects other unused ports and starts fine. I'm actually not unhappy with this behaviour - even though it differs from how Sparrow currently works,

Yup. This is why it is necessary to await this `AddressInfo` event being dispatched before configuring Http clients because if a port is taken. UX for lib consumers was awful if there was a failure due to a port being taken.

> how to determine the Tor control port in this situation - always read the file Tor writes, or is there a better way?

`TorManager` implements all the same interfaces that `TorController` does. It holds a reference to the controller and passes all [usecase interactions](https://github.com/05nelsonm/kmp-tor/tree/master/library/controller/kmp-tor-controller-common/src/commonMain/kotlin/io/matthewnelson/kmp/tor/controller/common/control/usecase) through right to the controller; the callback extension libs also accounts for these.

If you really do need access to the control port, yeah you can read the control port file. Here's how it's being done in `kmp-tor` on startup [link][(https://github.com/05nelsonm/kmp-tor/blob/1a847f532b57ba642d65366443080387bd48d0e6/library/manager/kmp-tor-manager/src/jvmAndroidMain/kotlin/io/matthewnelson/kmp/tor/manager/KmpTorLoader.kt#L325)

> Also, any ideas on another way to trigger failure to start?

Callback extension and use `torManager.start`. If tor fails to start in its initial phases (until bootstrap handoff), you will be returned the exception in the exception callback lambda.

As far as tor failing to bootstrap (sometimes tor gets stuck), that is the **only** scenario that is not handled by `kmp-tor` currently (see #47). In the meantime, you could create your own "bootstrap timeout" listener and attach it before calling `torManager.start`, then detach it when event `StartUpCompleteForTorInstance` gets dispatched. This particular issue's UX as far as library consumers is awful The current API structure of `kmp-tor` is a little rough, sorry. A `2.0.0` is underway to resolve a lot of the kinks that have been found through usage. Lots of changes for the better.

**Comment by 05nelsonm** [comment]
Sorry, didn't understand the q fully before answering.

>  Also, any ideas on another way to trigger failure to start?

If you want to trigger a failure to start, you can write a bad config option to the torrc file within your [provide](https://github.com/05nelsonm/kmp-tor/blob/1a847f532b57ba642d65366443080387bd48d0e6/library/manager/kmp-tor-manager/src/commonMain/kotlin/io/matthewnelson/kmp/tor/manager/TorConfigProvider.kt#L57) implementation. `provide` is called on a background thread (as noted in its documentation), so blocking there by writing to a file is totally fine.
